### ISSUE TITLE
Successful Training

### ISSUE THREAD

**Issue by jaydee829** [jrjbertram/jsbsim_rl#2]
I have been experimenting and looking through the codebase for gym-jsbsim and while I was not able to train DDPG on the turn to heading task, I was able to train it successfully on the constant heading task. Furthermore, I successfully trained PPO2 from stable baselines on the turn to heading task. I don't know enough about DDPG yet to diagnose how it might be implemented to learn turn to heading, but PPO2 required very little effort to use and I was able to parallelize it for much faster training times. Here's a graph of the rewards, DDPG in pink, PPO in blue.
![screenshot from 2019-02-22 11-23-20](https://user-images.githubusercontent.com/42184471/53255762-4e67dc80-3694-11e9-80d5-7a3dc29b4781.png)

**Comment by jrjbertram** [comment]
That looks like way better results than I was getting with the little bit I
played with it.  I should have some spare cycles this weekend to look into
the problem you found...  I apologize for not looking at it last week, but
I got totally slammed.  :oD

Couple of questions:
1) did you pull in the latest pull requests from the original gym-jsbsim
repo?  It looked like many bug fixes.

2) when you actually evaluate the flight either via plots or by running it
with flight gear, does it appear to be stable?  I was seeing for most of my
runs that it would seem to settle out with reward (I don't remember what
values), but when I hooked it up for evaluation it nose dived.  Most of the
time when I trained, it seemed to get into a state where it was choosing a
fixed action which resulted in something like pitching up and rolling to
the right, for example.

- Josh.

**Comment by jaydee829** [comment]
I am working on doing the evaluation now. I am getting errors whenever I try to run the eval script. I think it has to do with the interactive figure running in a console (I am getting the same errors as [https://stackoverflow.com/questions/42864809/tight-layout-throw-error-valueerror-max-arg-is-an-empty-sequence](url)

Unfortunately I am also getting these errors whether flightgear of human is selected as the render mode and I don't have enough knowledge of Docker to know what I might be doing wrong. I am currently running the eval from PyCharm. I'm going to try and figure out why this is happening, but if you've seen the problem or know how I might get around it, I'd appreciate the assist.

**Comment by jrjbertram** [comment]
Are you checking in on a fork somewhere?

**Comment by jaydee829** [comment]
No, I forked your codebase when I first started working with it, but I was just running your eval.py. I couldn't get the Dockerfile you provided to build so I copied the image from dockerhub and then have been using it to run everything from within PyCharm. I fixed the above issue by using a different backend for matplotlib (TkAGG), now I am running into this problem: `_tkinter.TclError: couldn't connect to display ":1"` I think I have seen similar errors when trying to make plots in a container, but since I can't get the Dockerfile to build, I can't really troubleshoot what I think the problem is.

**Comment by jaydee829** [comment]
Since I got the human rendering working I was able to look at what my agent had learned and it was as you had found, basically full aileron in one direction, then oscillating elevator with some rudder oscillation as well. If I had the flightgear rendering working it would probably just be traveling in a circle. I'm going to continue working with it, potentially try a different policy.

**Comment by jrjbertram** [comment]
Sounds like you're seeing about what I was seeing.  I have a spare hour or
two, I'll see if I can remember how to get the docker image to build.

**Comment by jaydee829** [comment]
Got the Dockerfile to build, like I said in the other thread, I needed to pull in a fresh download of the jsbsim code and a copy of stable baselines. The one thing I don't have working is flightgear. It may just be my lack of knowledge with that software. The gym code looks like it outputs to a UDP port which I assume flightgear uses, I guess I probably need to start flightgear first then, just haven't had the time to tinker with it. Most concerned with getting the policy to learn useful actions. There is a fork of gym_jsbsim that is relatively active, I might look at the changes they have made and see how that might affect things. I'm also going to check the rewards functions to understand those better.

**Comment by jrjbertram** [comment]
The ability to run flight gear was somewhat neutered in the gym_jsbsim I
forked from.  I ended up hacking in an environment variable.  If you set
the following environment variable before you run the evaluate script (or
train, I suppose), then this will force gym_jsbsim to use the flight gear
rendering path.

export GYM_JSBSIM_RENDER_MODE=flightgear

I can't remember if it you have to start flight gear yourself, or if it
starts its own copy.  (Pretty sure it starts its own with a call to fgfs in
the code.)

**Comment by jrjbertram** [comment]
https://gitter.im/jsbsim_rl/community

Created a gitter room as well if you prefer to ask questions there.  I'm good with either.

**Comment by beiliwzl** [comment]
> I have been experimenting and looking through the codebase for gym-jsbsim and while I was not able to train DDPG on the turn to heading task, I was able to train it successfully on the constant heading task. Furthermore, I successfully trained PPO2 from stable baselines on the turn to heading task. I don't know enough about DDPG yet to diagnose how it might be implemented to learn turn to heading, but PPO2 required very little effort to use and I was able to parallelize it for much faster training times. Here's a graph of the rewards, DDPG in pink, PPO in blue.
> ![screenshot from 2019-02-22 11-23-20](https://user-images.githubusercontent.com/42184471/53255762-4e67dc80-3694-11e9-80d5-7a3dc29b4781.png)

hi, i am also working on it recently, but i can not tran the agent sucessfully like your results, where the average reward can reach above 60. Can you share your hyper parameters with me, or some tricks need to attation, thank you very much.
